markets/perps-market/contracts/modules/AsyncOrderCancelModule.sol

```diff
@@ -0,0 +1,237 @@
+    using Position for Position.Data;
+    using SafeCastU256 for uint256;
+    using SafeCastI256 for int256;
+
+    int256 private constant PRECISION = 18;
+
+    /**
+     * @inheritdoc IAsyncOrderCancelModule
+     */
+    function cancelOrder(uint128 accountId) external view {
+        GlobalPerpsMarket.load().checkLiquidation(accountId);
+        (
+            AsyncOrder.Data storage order,
+            SettlementStrategy.Data storage settlementStrategy
+        ) = AsyncOrder.loadValid(accountId);
+
+        _cancelOffchain(order, settlementStrategy);
+    }
+
+    /**
+     * @inheritdoc IAsyncOrderCancelModule
+     */
+    function cancelPythOrder(bytes calldata result, bytes calldata extraData) external payable {
+        (
+            uint256 offchainPrice,
+            AsyncOrder.Data storage order,
+            SettlementStrategy.Data storage settlementStrategy
+        ) = _parsePythPrice(result, extraData);
+
+        _cancelOrder(offchainPrice, order, settlementStrategy);
+    }
+
+    /**
+     * @dev used for canceling offchain orders. This will revert with OffchainLookup.
+     */
+    function _cancelOffchain(
+        AsyncOrder.Data storage asyncOrder,
+        SettlementStrategy.Data storage settlementStrategy
+    ) private view returns (uint, int256, uint256) {
+        string[] memory urls = new string[](1);
+        urls[0] = settlementStrategy.url;
+
+        bytes4 selector;
+        if (settlementStrategy.strategyType == SettlementStrategy.Type.PYTH) {
+            selector = AsyncOrderCancelModule.cancelPythOrder.selector;
+        } else {
+            revert SettlementStrategyNotFound(settlementStrategy.strategyType);
+        }
+
+        // see EIP-3668: https://eips.ethereum.org/EIPS/eip-3668
+        revert OffchainLookup(
+            address(this),
+            urls,
+            abi.encodePacked(settlementStrategy.feedId, _getTimeInBytes(asyncOrder.settlementTime)),
+            selector,
+            abi.encode(asyncOrder.request.accountId) // extraData that gets sent to callback for validation
+        );
+    }
+
+    /**
+     * @dev used for canceling an order.
+     */
+    function _cancelOrder(
+        uint256 price,
+        AsyncOrder.Data storage asyncOrder,
+        SettlementStrategy.Data storage settlementStrategy
+    ) private {
+        CancelOrderRuntime memory runtime;
+        runtime.marketId = asyncOrder.request.marketId;
+        runtime.accountId = asyncOrder.request.accountId;
+        runtime.acceptablePrice = asyncOrder.request.acceptablePrice;
+        runtime.settlementReward = settlementStrategy.settlementReward;
+        runtime.sizeDelta = asyncOrder.request.sizeDelta;
+
+        // check if account is flagged
+        GlobalPerpsMarket.load().checkLiquidation(runtime.accountId);
+
+        // Validate Order still exists and is valid
+        if (runtime.sizeDelta == 0) {
+            revert AsyncOrder.ZeroSizeOrder();
+        }
```

Comment:
i think this and the check for within settlement window can be grouped into a `asyncOrder.validateCancellation()` function.  

- there has to be a pending order
- and the order needs to be within settlement window

Reply:
as mentioned before, the settlement window was checked when parsing the pyth price some function calls above